### PR TITLE
community: OpenAPI agent. Add support for PUT, DELETE and PATCH

### PR DESCRIPTION
**Pull request by pow3rpi**
**Description**: Add PUT, DELETE and PATCH tools to tool list for OpenAPI agent if dangerous requests are allowed.

**Issue**: https://github.com/langchain-ai/langchain/issues/20469

### CODE REVIEW

**Review by eyurtsev — CHANGES_REQUESTED** [review]
This is a breaking change that adds PUT, DELETE and PATCH capabilities to an existing model. If you'd like to add it needs to be opt-in and likely controllable as to which operations should be included.

This could be a good way to parameterize:

```
allow_operations: Sequence[str] = ('get', 'post', ...) # With default matching current defaults
````

**Comment by pow3rpi** [comment]
@eyurtsev Ok, it is a good point, I will add it.

**Comment by pow3rpi** [comment]
@eyurtsev GET requests will be definitely in the default settings but do you want to keep POST requests there as well?

**Comment by pow3rpi** [comment]
@eyurtsev If you are ok with the update, I believe we should also update this [doc](https://github.com/langchain-ai/langchain/blob/master/docs/docs/integrations/toolkits/openapi.ipynb).

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py [hunk 1]
```diff
@@ -346,6 +378,7 @@ def create_openapi_agent(
     verbose: bool = True,
     agent_executor_kwargs: Optional[Dict[str, Any]] = None,
     allow_dangerous_requests: bool = False,
+    allow_operations: Sequence[str] = ('get', 'post'),
```
Comment (eyurtsev) [inline libs/community/langchain_community/agent_toolkits/openapi/planner.py:381]: This looks good.

Could you introduce a type for this so users don't have to guess what the set of supported operations is?

```python
Operation = Literal['GET', 'POST', 'DELETE', 'PUT', 'PATCH']
```

```suggestion
    allow_operations: Sequence[Operation] = ('GET', 'POST'),
```

**Review by eyurtsev — COMMENTED** [review]
@pow3rpi Would you be able to document the type? we can then merge this change

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py (hunk 1 above)
Comment (pow3rpi) [reply libs/community/langchain_community/agent_toolkits/openapi/planner.py:381]: @eyurtsev I've added Literals, so could you please explain where exactly I have to document them? Should I add it here in [docs](https://github.com/langchain-ai/langchain/blob/master/docs/docs/integrations/toolkits/openapi.ipynb) or anywhere else?

**Comment by pow3rpi** [comment]
@eyurtsev Also, could you guess why [dependency checks](https://github.com/langchain-ai/langchain/actions/runs/9581469727/job/26486098898?pr=22962) and [tests](https://github.com/langchain-ai/langchain/actions/runs/9581469727/job/26486097518?pr=22962) fail with the following error?

> =========================== short test summary info ============================
> FAILED tests/unit_tests/embeddings/test_ovhcloud.py::test_ovhcloud_embed_documents - ValueError: Request failed with status code: 401, {"message":"Bad token; invalid JSON"}
> ========== 1 failed, 890 passed, 471 skipped, 4440 warnings in 11.71s ==========
> make: *** [Makefile:19: test] Error 1
> Error: Process completed with exit code 2.

**Comment by eyurtsev** [comment]
It's a bug on master. I'm on mobile right now so can't do it from the phone, but you should be able to use the UI to update the PR to use latest master, if not I'll do it on your behalf when back at a computer

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py [hunk 2]
```diff
@@ -346,6 +385,7 @@ def create_openapi_agent(
     verbose: bool = True,
     agent_executor_kwargs: Optional[Dict[str, Any]] = None,
     allow_dangerous_requests: bool = False,
+    allow_operations: Sequence[Operation] = ('GET', 'POST'),
```
Comment (eyurtsev) [inline libs/community/langchain_community/agent_toolkits/openapi/planner.py:388]:
```suggestion
    allowed_operations: Sequence[Operation] = ('GET', 'POST'),
```

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py [hunk 3]
```diff
@@ -371,7 +411,7 @@ def create_openapi_agent(
     tools = [
         _create_api_planner_tool(api_spec, llm),
         _create_api_controller_tool(
-            api_spec, requests_wrapper, llm, allow_dangerous_requests
+            api_spec, requests_wrapper, llm, allow_dangerous_requests, allow_operations
```
Comment (eyurtsev) [inline libs/community/langchain_community/agent_toolkits/openapi/planner.py:414]:
```suggestion
            api_spec, requests_wrapper, llm, allow_dangerous_requests, allowed_operations
```

**Review by eyurtsev — APPROVED** [review]

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py (hunk 1 above)
Comment (pow3rpi) [reply libs/community/langchain_community/agent_toolkits/openapi/planner.py]: @eyurtsev I saw your remarks below but what you can say about documentation? I guess you missed my previous message in this unresolved conversation?

**Comment by eyurtsev** [comment]
@pow3rpi could you allow maintainers to edit the PR -- that would allow me to resolve issues and merge in the future

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py (hunk 3 above)
Comment (pow3rpi) [reply libs/community/langchain_community/agent_toolkits/openapi/planner.py]: @eyurtsev  Changes have been made to address this issue. Please see [commit](https://github.com/langchain-ai/langchain/pull/22962/commits/762f03d17bb5d55a390b101b4ebda160d8933394).

File: libs/community/langchain_community/agent_toolkits/openapi/planner.py (hunk 2 above)
Comment (pow3rpi) [reply libs/community/langchain_community/agent_toolkits/openapi/planner.py]: @eyurtsev  Changes have been made to address this issue. Please see [commit](https://github.com/langchain-ai/langchain/pull/22962/commits/762f03d17bb5d55a390b101b4ebda160d8933394).

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**